domain-models-runtime/src/main/resources/templates/db_scripts/optimistic_locking.ftl

```diff
@@ -0,0 +1,49 @@
+-- auto update optimistic locking version
+
+<#if table.withOptimisticLocking??>
+  <#assign ol_version = "_version">
+  <#assign ol_notice_level = "WARNING">
+  CREATE OR REPLACE FUNCTION ${myuniversity}_${mymodule}.${table.tableName}_set_ol_version()
+  RETURNS trigger AS $$
+  BEGIN
+    CASE TG_OP
+      WHEN 'INSERT' THEN
+        <#if table.withOptimisticLocking.name() == "OFF">
+          NEW.jsonb = NEW.jsonb - '${ol_version}';
+        <#else>
+          NEW.jsonb = jsonb_set(NEW.jsonb, '{${ol_version}}', to_jsonb(1));
+        </#if>
+      WHEN 'UPDATE' THEN
+        <#if table.withOptimisticLocking.name() == "OFF">
+          NEW.jsonb = NEW.jsonb - '${ol_version}';
+        <#else>
+          IF NEW.jsonb->'${ol_version}' IS DISTINCT FROM OLD.jsonb->'${ol_version}' THEN
+            <#if table.withOptimisticLocking.name() == "FAIL">
+              <#assign ol_notice_level = "EXCEPTION">
+            </#if>
+            RAISE ${ol_notice_level} 'Cannot update record % because it has been changed: '
```

Comment:
Can you also set some error code, for example 23F09 (23 for class 23 — Integrity Constraint Violation, F for FOLIO, 09 for 409 HTTP status code), to make it more easy to detect this error or warning?
https://www.postgresql.org/docs/current/errcodes-appendix.html

Reply:
> Can you also set some error code, for example 23F09 (23 for class 23 — Integrity Constraint Violation, F for FOLIO, 09 for 409 HTTP status code), to make it more easy to detect this error or warning?
> https://www.postgresql.org/docs/current/errcodes-appendix.html

Good suggestion. Done.